### PR TITLE
[ca] attach compiled autograd to its compile id in tlparse

### PR DESCRIPTION
**Pull request by xmfan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #141408
* __->__ #141362
* #140964

This delays the tlparse for compiled autograd into the dynamo call for this corresponding compiled autograd graph. Not perfect, but is an improvement over not having the graphs bundled with their torch.compile artifacts.

It means a new file will be logged for the same CA graph every time dynamo recompiles. Currently to tell if you have a recompile or not, you'd have to click into the graph and look at the name of graph e.g. CompiledAutograd0 vs CompiledAutograd1. I'm thinking of also adding the CA id to the tlparse file name.

tlparse: https://interncache-all.fbcdn.net/manifold/tlparse_reports/tree/logs/.tmpKvl4hk/index.html

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @yf225

### CODE REVIEW

**Comment by ezyang** [comment]
@masnesral weren't you and @xmfan talking about how to adjust compile id for compiled autograd?

Concretely, I think compiled autograd can merit it's own id which is a PREFIX to the regular compile id. This prefix is per compiled autograd graph. Then for inner graphs they can just get a fresh compile id (or many, since you may have many graph breaks).

I don't think delaying is good, because (1) it means if we crash before we get to Dynamo the output is lost forever and (2) you are attributing compiled autograd trace to the FIRST dynamo graph inside the CA graph, but really it has nothing to do with the first graph, it's  ALL of the graphs.

**Comment by masnesral** [comment]
> weren't you and @xmfan talking about how to adjust compile id for compiled autograd?

I'll let @xmfan explain. But TL;DR I think was worried about anything that would change the format of the compile_id since he didn't have a good handle on what depends on the current format (nor do I)

**Comment by ezyang** [comment]
I feel I have kept reasonably good tabs on this, and I haven't seen any code splitting on the slash to extract things out (except maybe Scuba queries, but that's fine). I think it's worth it. There's some other things I wanted to change (mostly distinguishing between compiles of different nn modules) which we can hit while we're at it.

**Comment by xmfan** [comment]
@ezyang The end state of compiled autograd integration should be single step dynamo, and removing the graph break on `loss.backward()`. So a separate compile id for compiled autograd is only needed interim.

**Comment by xmfan** [comment]
> I don't think delaying is good, because (1) it means if we crash before we get to Dynamo the output is lost forever and (2) you are attributing compiled autograd trace to the FIRST dynamo graph inside the CA graph, but really it has nothing to do with the first graph, it's ALL of the graphs. 

I think there's very little risk to (1): before the PR, we were logging at the end of compiled autograd, and if we crash in that region, we need to look at stack traces in logs, no tlparse artifacts would be useful. 

Under what cases would (2) be true? It still work for tested graph breaks scenarios. Overall, it's still an improvement over what we have today.

**Comment by ezyang** [comment]
> So a separate compile id for compiled autograd is only needed interim.

That's not how instrumentation works. The point of instrumentation is to describe the system as it exists today. You don't... not instrument something just because you think it is going away (at some unspecified point in the future.)

> I think there's very little risk to (1): before the PR, we were logging at the end of compiled autograd, and if we crash in that region, we need to look at stack traces in logs, tlparse artifacts wouldn't be super useful.

Well, if I crash after I've finished constructing the compiled autograd graph, but before I actually get to running Dynamo, MAN, I *suuuure* would like to have known what the constructed compiled autograd graph is, no?
